subprojects/build-events/src/main/java/org/gradle/internal/build/event/DefaultBuildEventsListenerRegistry.java

```diff
@@ -136,22 +147,27 @@ private void processIfBuildService(Provider<?> listenerProvider) {
     }
 
     private void unsubscribeProvider(Provider<?> listenerProvider) {
-        AbstractListener<?> subscription = subscriptions.remove(listenerProvider);
+        AbstractListener<?> subscription;
+        synchronized (subscriptions) {
+            subscription = subscriptions.remove(listenerProvider);
+        }
         if (subscription != null) {
             subscription.getListeners().forEach(this::unsubscribe);
             subscription.close();
         }
     }
 
     private void unsubscribeAll() {
-        try {
-            Collection<AbstractListener<?>> subscribed = subscriptions.values();
-            subscribed.stream()
-                .flatMap(it -> it.getListeners().stream())
-                .forEach(this::unsubscribe);
-            CompositeStoppable.stoppable(subscribed).stop();
-        } finally {
-            subscriptions.clear();
+        synchronized (subscriptions) {
+            try {
+                Collection<AbstractListener<?>> subscribed = subscriptions.values();
+                subscribed.stream()
+                    .flatMap(it -> it.getListeners().stream())
+                    .forEach(this::unsubscribe);
+                CompositeStoppable.stoppable(subscribed).stop();
```

Comment:
🤔 Wouldn't it be safer to stop the subscribers outside the lock just in case they try to reenter by accident?

Reply:
> 🤔 Wouldn't it be safer to stop the subscribers outside the lock just in case they try to reenter by accident?

Given that we're waiting for a thread running user code in `stop()`, some safeguards make sense, thanks. I should have followed my habits :). 

Ideally, we should handle when something adds a listener after cleanup, but this would be a breaking change, I'm reluctant to adding this in a patch release - it needs a deprecation warning first.